src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6358,10 +6454,219 @@ MutationCommands StorageReplicatedMergeTree::getFirstAlterMutationCommandsForPar
+
+    LOG_TRACE(log, "Set zookeeper lock {}", zookeeper_node);
+
+    /// In rare case other replica can remove path between createAncestors and createIfNotExists
+    /// So we make up to 5 attempts
+    for (int attempts = 5; attempts > 0; --attempts)
+    {
+        try
+        {
+            zookeeper->createAncestors(zookeeper_node);
+            zookeeper->createIfNotExists(zookeeper_node, "lock");
+            break;
+        }
+        catch (const zkutil::KeeperException & e)
+        {
+            if (e.code == Coordination::Error::ZNONODE)
+                continue;
+            throw;
+        }
+    }
+}
+
+
+bool StorageReplicatedMergeTree::unlockSharedData(const IMergeTreeDataPart & part) const
+{
+    return unlockSharedData(part, part.relative_path);
+}
+
+
+bool StorageReplicatedMergeTree::unlockSharedData(const IMergeTreeDataPart & part, const String & path) const
+{
+    if (!part.volume)
+        return true;
+    DiskPtr disk = part.volume->getDisk();
+    if (!disk)
+        return true;
+    if (disk->getType() != DB::DiskType::Type::S3)
+        return true;
+
+    zkutil::ZooKeeperPtr zookeeper = tryGetZooKeeper();
+    if (!zookeeper)
+        return true;
+
+    String id = part.getUniqueId();
+    boost::replace_all(id, "/", "_");
+    String norm_path = path;
+    boost::replace_all(norm_path, "/", "_");
+
+    String zookeeper_part_node = zookeeper_path + "/zero_copy_s3/shared/" + part.name;
+    String zookeeper_part_uniq_node = zookeeper_part_node + "/" + id;
+    String zookeeper_node = zookeeper_part_uniq_node + "/" + replica_name;
+
+    LOG_TRACE(log, "Remove zookeeper lock {}", zookeeper_node);
+
+    zookeeper->tryRemove(zookeeper_node);
+
+    Strings children;
+    zookeeper->tryGetChildren(zookeeper_part_uniq_node, children);
+
+    if (!children.empty())
+    {
+        LOG_TRACE(log, "Found zookeper locks for {}", zookeeper_part_uniq_node);
+        return false;
+    }
+
+    zookeeper->tryRemove(zookeeper_part_uniq_node);
+
+    /// Even when we have lock with same part name, but with different uniq, we can remove files on S3
+    children.clear();
+    zookeeper->tryGetChildren(zookeeper_part_node, children);
+    if (children.empty())
+        /// Cleanup after last uniq removing
+        zookeeper->tryRemove(zookeeper_part_node);
+
+    return true;
+}
+
+
+bool StorageReplicatedMergeTree::tryToFetchIfShared(
+    const IMergeTreeDataPart & part,
+    const DiskPtr & disk,
+    const String & path) const
+{
+    const auto data_settings = getSettings();
+    if (!data_settings->allow_s3_zero_copy_replication)
+        return false;
+
+    if (disk->getType() != DB::DiskType::Type::S3)
+        return false;
+
+    String replica = getSharedDataReplica(part);
+
+    /// We can't fetch part when none replicas have this part on S3
+    if (replica.empty())
+        return false;
+
+    ReplicatedMergeTreeLogEntry log_entry;
+    log_entry.type = ReplicatedMergeTreeLogEntry::FETCH_SHARED_PART;
```

Comment:
Seems like there are no places where we use it? Never store in replication log? Maybe we don't need it?